### PR TITLE
refactor(log): replace panics with log fatal or log panic functions

### LINKED ISSUE #1686: [Refactor] Zot should not use panics, it should use log.Fatal
We need to closely review all panics used in the zot code, and evaluate replacing them with log.Fatal.
This may or may not make sense on a case by case basis (maybe the logger is not initialized, or maybe this is zli code not zot code), but we should review all these cases nevertheless.

### PR DESCRIPTION
**Pull request by laurentiuNiculae**
**What type of PR is this?**

**Which issue does this PR fix**:
Closes #1686

**What does this PR do / Why do we need it**:

**If an issue # is not available please add repro steps and logs showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this break upgrades or downgrades?**

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/api/authn.go [hunk 1]
```diff
@@ -548,19 +553,20 @@ func (rh *RouteHandler) AuthURLHandler() http.HandlerFunc {
...
-func NewRelyingPartyOIDC(config *config.Config, provider string) rp.RelyingParty {
-	issuer, clientID, clientSecret, redirectURI, scopes, options := getRelyingPartyArgs(config, provider)
+func NewRelyingPartyOIDC(config *config.Config, provider string, log log.Logger) rp.RelyingParty {
+	issuer, clientID, clientSecret, redirectURI, scopes, options := getRelyingPartyArgs(config, provider, log)
 
 	relyingParty, err := rp.NewRelyingPartyOIDC(issuer, clientID, clientSecret, redirectURI, scopes, options...)
 	if err != nil {
-		panic(err)
+		log.Panic().Err(err).Str("issuer", issuer).Str("redirectURI", redirectURI).Strs("scopes", scopes).
+			Msg("failed to get new rekying party oicd")
```
Comment (andaaron) [inline pkg/api/authn.go:562]:
```suggestion
			Msg("failed to get new relying party oicd")
```

File: pkg/api/authn.go [hunk 2]
```diff
@@ -572,17 +578,18 @@ func NewRelyingPartyGithub(config *config.Config, provider string) rp.RelyingPar
 
 	relyingParty, err := rp.NewRelyingPartyOAuth(rpConfig, options...)
 	if err != nil {
-		panic(err)
+		log.Panic().Err(err).Str("redirectURI", redirectURI).Strs("scopes", scopes).
+			Msg("failed to get new rekying party oauth")
```
Comment (andaaron) [inline pkg/api/authn.go:582]:
```suggestion
			Msg("failed to get new relying party oauth")
```

File: pkg/api/authn.go [hunk 3]
```diff
@@ -315,7 +319,8 @@ func (amw *AuthnMiddleware) TryAuthnHandlers(ctlr *Controller) mux.MiddlewareFun
 	if ctlr.Config.IsHtpasswdAuthEnabled() {
 		credsFile, err := os.Open(ctlr.Config.HTTP.Auth.HTPasswd.Path)
 		if err != nil {
-			panic(err)
+			amw.log.Fatal().Err(err).Str("credsFile", ctlr.Config.HTTP.Auth.HTPasswd.Path).
```
Comment (andaaron) [inline pkg/api/authn.go:322]: Please don't use Fatal() log messages, as deferred code would not be executed.

File: pkg/meta/meta.go [hunk 4]
```diff
@@ -49,7 +49,7 @@ func Create(dbtype string, dbDriver, parameters interface{}, log log.Logger, //n
 		{
 			properDriver, ok := dbDriver.(*bbolt.DB)
 			if !ok {
-				panic("failed type assertion")
+				log.Panic().Msg("failed type assertion")
```
Comment (andaaron) [inline pkg/meta/meta.go:52]: Return the error, instead of panic-ing

File: pkg/storage/cache/boltdb.go [hunk 5]
```diff
@@ -30,7 +30,7 @@ type BoltDBDriverParameters struct {
 func NewBoltDBCache(parameters interface{}, log zlog.Logger) Cache {
 	properParameters, ok := parameters.(BoltDBDriverParameters)
 	if !ok {
-		panic("Failed type assertion")
+		log.Panic().Msg("Failed type assertion")
```
Comment (andaaron) [inline pkg/storage/cache/boltdb.go:33]: Can you try to log the error and return nil?

File: pkg/storage/cache/dynamodb.go [hunk 6]
```diff
@@ -64,7 +64,7 @@ func (d *DynamoDBDriver) NewTable(tableName string) error {
 func NewDynamoDBCache(parameters interface{}, log zlog.Logger) Cache {
 	properParameters, ok := parameters.(DynamoDBDriverParameters)
 	if !ok {
-		panic("Failed type assertion!")
+		log.Panic().Msg("Failed type assertion")
```
Comment (andaaron) [inline pkg/storage/cache/dynamodb.go:67]: Can you try to log the error and return nil?

File: pkg/api/controller.go [hunk 7]
```diff
@@ -188,13 +188,13 @@ func (c *Controller) Run(reloadCtx context.Context) error {
 
 			caCert, err := os.ReadFile(c.Config.HTTP.TLS.CACert)
 			if err != nil {
-				panic(err)
+				c.Log.Panic().Err(err).Str("caCert", c.Config.HTTP.TLS.CACert).Msg("failed to read file")
```
Comment (andaaron) [inline pkg/api/controller.go:191]: Can we return the error instead?

File: pkg/storage/cache.go [hunk 8]
```diff
@@ -46,27 +46,31 @@ func CreateCacheDatabaseDriver(storageConfig config.StorageConfig, log zlog.Logg
...
 }
 
 func Create(dbtype string, parameters interface{}, log zlog.Logger) (cache.Cache, error) {
 	switch dbtype {
 	case "boltdb":
 		{
-			return cache.NewBoltDBCache(parameters, log), nil
+			cache, err := cache.NewBoltDBCache(parameters, log)
+
+			if errors.Is(err, zerr.ErrTimeout) {
```
Comment (andaaron) [inline pkg/storage/cache.go:61]: Why is this needed?

File: pkg/api/controller_test.go [hunk 9]
```diff
@@ -1297,32 +1305,38 @@ func TestMultipleInstance(t *testing.T) {
...
-		resp, _ = resty.R().SetBasicAuth(username, passphrase).Get(baseURL)
-		So(resp, ShouldNotBeNil)
-		So(resp.StatusCode(), ShouldEqual, http.StatusNotFound)
+		err = ctlr.Init(context.Background())
+		So(err, ShouldNotBeNil)
 
-		resp, _ = resty.R().SetBasicAuth(username, passphrase).Get(baseURL + "/v2/")
-		So(resp, ShouldNotBeNil)
-		So(resp.StatusCode(), ShouldEqual, http.StatusOK)
+		// T O D O Delete after you make sure this is the right behavior
```
Comment (andaaron) [inline pkg/api/controller_test.go:1318]: Why do we have doubts here?

File: pkg/storage/cache.go [hunk 10]
```diff
@@ -46,27 +44,31 @@ func CreateCacheDatabaseDriver(storageConfig config.StorageConfig, log zlog.Logg
...
 }
 
 func Create(dbtype string, parameters interface{}, log zlog.Logger) (cache.Cache, error) {
 	switch dbtype {
 	case "boltdb":
 		{
-			return cache.NewBoltDBCache(parameters, log), nil
+			cache, err := cache.NewBoltDBCache(parameters, log)
+
+			// if errors.Is(err, zerr.ErrTimeout) {
```
Comment (andaaron) [inline pkg/storage/cache.go:59]: Please uncomment or remove the code.

**Review by andaaron — APPROVED** [review]

**Comment by rchincha** [comment]
codecov/project Failing after 1s — 91.49% (-0.24%) compared to a11fe2d

File: pkg/storage/cache/dynamodb.go [hunk 11]
```diff
@@ -85,18 +88,20 @@ func NewDynamoDBCache(parameters interface{}, log zlog.Logger) Cache {
 	if err != nil {
 		log.Error().Err(err).Msg("unable to load AWS SDK config for dynamodb")
 
-		return nil
+		return nil, err
```
Comment (andaaron) [inline pkg/storage/cache/dynamodb.go:91]: Maybe we can pass an invalid config in the environment for this test? But make sure it does not stay in the environment.

File: pkg/api/controller_test.go [hunk 12]
```diff
@@ -1266,7 +1273,7 @@ func TestMultipleInstance(t *testing.T) {
 
 	Convey("Test zot multiple subpath with same root directory", t, func() {
 		port := test.GetFreePort()
-		baseURL := test.GetBaseURL(port)
+		// baseURL := test.GetBaseURL(port)
```
[the remark this answers is not in the document]
Comment (andaaron) [reply pkg/api/controller_test.go:1290]: This comment still applies.

File: pkg/storage/storage_test.go [hunk 13]
```diff
@@ -871,6 +883,73 @@ func TestMandatoryAnnotations(t *testing.T) {
...
+
+	Convey("Create unique subpath, error for cache driver", t, func() {
+		tmpDirSubpath := t.TempDir()
+		config := &config.Config{
+			Storage: config.GlobalStorageConfig{
+				StorageConfig: config.StorageConfig{RootDirectory: t.TempDir()},
+				SubPaths: map[string]config.StorageConfig{
+					"a/": {
+						RootDirectory: tmpDirSubpath,
+						// RemoteCache:   true,
```
Comment (andaaron) [inline pkg/storage/storage_test.go:911]: Why is this commented out?

**Review by andaaron — DISMISSED** [review]

**Review by andaaron — APPROVED** [review]

**Review by andaaron — APPROVED** [review]
